src/Symfony/Component/DependencyInjection/Compiler/AttributeAutoconfigurationPass.php

```diff
@@ -11,47 +11,103 @@
 
 /**
  * @author Alexander M. Turek <me@derrabus.de>
  */
-final class AttributeAutoconfigurationPass implements CompilerPassInterface
+final class AttributeAutoconfigurationPass extends AbstractRecursivePass
 {
     private $ignoreAttributesTag;
+    private $argumentConfigurators;
 
     public function __construct(string $ignoreAttributesTag = 'container.ignore_attributes')
     {
         $this->ignoreAttributesTag = $ignoreAttributesTag;
+
+        $this->argumentConfigurators = [
+            BindTaggedIterator::class => static function (BindTaggedIterator $attribute) {
+                return new TaggedIteratorArgument($attribute->tag, $attribute->indexAttribute);
+            },
+            BindTaggedLocator::class => static function (BindTaggedLocator $attribute) {
+                return new ServiceLocatorArgument(new TaggedIteratorArgument($attribute->tag, $attribute->indexAttribute));
+            },
+        ];
     }
 
     public function process(ContainerBuilder $container): void
     {
-        if (80000 > \PHP_VERSION_ID) {
-            return;
+        if (80000 <= \PHP_VERSION_ID) {
+            parent::process($container);
+        }
+    }
+
+    protected function processValue($value, bool $isRoot = false)
+    {
+        if ($value instanceof Definition
+            && $value->isAutoconfigured()
+            && !$value->isAbstract()
+            && !$value->hasTag($this->ignoreAttributesTag)
+        ) {
+            $value = $this->processDefinition($value);
+        }
+
+        return parent::processValue($value, $isRoot);
+    }
+
+    private function processDefinition(Definition $definition): Definition
+    {
+        if (!$reflector = $this->container->getReflectionClass($definition->getClass(), false)) {
+            return $definition;
         }
 
-        $autoconfiguredAttributes = $container->getAutoconfiguredAttributes();
+        $autoconfiguredAttributes = $this->container->getAutoconfiguredAttributes();
 
-        foreach ($container->getDefinitions() as $id => $definition) {
-            if (!$definition->isAutoconfigured()
-                || $definition->isAbstract()
-                || $definition->hasTag($this->ignoreAttributesTag)
-                || !($reflector = $container->getReflectionClass($definition->getClass(), false))
-            ) {
-                continue;
+        $instanceof = $definition->getInstanceofConditionals();
+        $conditionals = $instanceof[$reflector->getName()] ?? new ChildDefinition('');
+        foreach ($reflector->getAttributes() as $attribute) {
+            if ($configurator = $autoconfiguredAttributes[$attribute->getName()] ?? null) {
+                $configurator($conditionals, $attribute->newInstance(), $reflector);
             }
+        }
+
+        if ($constructor = $this->getConstructor($definition, false)) {
+            $definition = $this->bindArguments($definition, $constructor);
```

Comment:
what about setters?
it's no easy because we might not know them all yet (eg `@required` ones)
I think this means we should split this pass in two, so that we could register both concerns at separate steps

Reply:
I wasn't sure if we should really browse all public methods for potential setters. I think this feature is good enough if we limit it to constructor arguments, what do you think?